Add Prediction Guard Components (Running on Gaudi in ITDC)

dwhitena opened the pull request:
## Description

This PR does the following: 
1. Integrates [Prediction Guard](https://predictionguard.com/) powered OPEA components for LLM, LVM, embeddings, and guardrails. Prediction Guard hosts open access models on top of Gaudi in Intel Tiber Developer Cloud (ITDC). Users of Prediction Guard can use the models via a multi-tenant system (and an API key) or single-tenant/self-hosted, which provides a pathway for OPEA users to have turn-key model hosting of open models on top of Gaudi. That is, OPEA users can utilize OPEA to spin up OPEA compliant components with models hosted in Prediction Guard, which unlocks the various tooling and examples that will be built in an OPEA compliant ways.
2. Organizes the Guardrails component directory to allow for multiple providers of guardrails of various types. Rather than having a single way to run a single type of guardrail. The directory is now organized into directories for each type of guardrail (factuality, pii, etc.) with compatible components (from potentially different providers) underneath.
3. Organizes the LVM component directory to allow for multiple providers of LVMs. Currently there is the option to run the LVM component via transformers and Prediction Guard.
4. Fixes various typos or minor README clarifications as they were found.

We wanted to pull in all of these together to make sure that they all can work together as OPEA compliant components in the GenAIExamples. We tested each service individually and orchestrated together. 

## Issues

n/a

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [x] Others (enhancement, documentation, validation, etc.)

## Dependencies

The Prediction Guard Python SDK:
- [Docs](https://docs.predictionguard.com/docs/getting-started/welcome)
- [Github](https://github.com/predictionguard/python-client)
- [PyPI](https://pypi.org/project/predictionguard/)

## Tests

All of the deployment and cURL commands listed in all respective directories were run by multiple people on both Linux and Mac machines to validate the responses from the services.

Spycsh commented on comps/cores/proto/docarray.py:
```diff
@@ -62,6 +62,10 @@ class EmbedDoc1024(BaseDoc):
     text: str
     embedding: conlist(float, min_length=1024, max_length=1024)
 
+class EmbedDoc512(BaseDoc):
```
Is this length 512 a must? We actually plan to remove all the hard limitation of embedding length in this [PR](https://github.com/opea-project/GenAIComps/pull/347). Do you think that fit this case?

dwhitena replied:
The embeddings we will return from our current embedding model are 512, but there isn't a need to have this as a strict constraint in the pydandic model. I've removed the constraints for now and added a `TODO` comment to update to the new `EmbedDoc` whenever that is merged (depending if that comes before or after this PR is merged.

dwhitena replied:
Ah @Spycsh, I didn't realize that other PR was already merged. I updated our embedding component accordingly.

Spycsh commented on comps/lvms/README.md:
````diff
@@ -1,111 +1,18 @@
-```bash
-pip install -r requirements.txt
-```
+The [transformers](transformers) directory contains instructions for running services that serve predictions from a LLaVA LVM. Two services must be spun up to run the LVM:
````
I think you should also delete all the content under the legacy folder https://github.com/predictionguard/pg-GenAIComps/tree/main/comps/lvms/llava, since they are moved to `transformers`. Let me know if I miss something!

dwhitena replied:
Ah, good catch. Yes, that was the intention, and I've deleted the unnecessary files.

Spycsh commented:
Hi @dwhitena , this Prediction Guard component seems powerful and well-organized! I have just had a few comments above. Please let me know if you have further questions!

dwhitena commented:
Talked with @sharanshirodkar7 about the changes here. He was able to run the pre-commit locally to resolve the workflow checks, but that also introduced a good number of style etc. changes throughout. Hence the large number of files changed. Let us know if this is an issue.

dwhitena commented:
> Talked with @sharanshirodkar7 about the changes here. He was able to run the pre-commit locally to resolve the workflow checks, but that also introduced a good number of style etc. changes throughout. Hence the large number of files changed. Let us know if this is an issue.

Ah, @sharanshirodkar7 caught some virtual env stuff that got added and cleaned that up. Thanks @sharanshirodkar7

sharanshirodkar7 commented:
Yes. Added "venv/" to the .gitignore file too so that this doesn't happen again.

lvliang-intel commented:
@dwhitena, 
Please resolve the code conflict.

XuhuiRen commented on comps/dataprep/redis/langchain/prepare_doc_redis.py:
```diff
@@ -10,6 +10,7 @@
 
 # from pyspark import SparkConf, SparkContext
 import redis
+import uvicorn
```
why do you need to import this package in this file but do not use it

XuhuiRen commented on comps/embeddings/predictionguard/README.md:
```diff
@@ -0,0 +1,39 @@
+
+[Prediction Guard](https://docs.predictionguard.com) allows you to utilize hosted open access LLMs, LVMs, and embedding functionality with seamlessly integrated safeguards. In addition to providing a scalable access to open models, Prediction Guard allows you to configure factual consistency checks, toxicity filters, PII filters, and prompt injection blocking. Join the [Prediction Guard Discord c ...
+
+This embedding microservice is designed to efficiently convert text into vectorized embeddings using the [BridgeTower model](https://huggingface.co/BridgeTower/bridgetower-large-itm-mlm-itc). Thus, it is ideal for both RAG or semantic search applications.
```
could be better if you can add some introduction about the relation between embedding service and the predictionguard.

XuhuiRen commented:
please add UTs for checking the code functionalities

sharanshirodkar7 commented:
@XuhuiRen We have resolved the conflicts, added UTs to all our PG microservices and tested its functionality. Please approve.

ftian1 commented on comps/llms/text-generation/predictionguard/README.md:
```diff
@@ -0,0 +1,54 @@
+# Introduction
```
shall we enable this by default but not introduce a new instance under llms/text-generation?

XuhuiRen reviewed (approved):
Please check failed UTs

chensuyue commented:
https://github.com/opea-project/GenAIComps/actions/runs/10582938136/job/29344664641?pr=344#step:4:57
This issue means, the changed path `comps/guardrails/pii_detection/docker/Dockerfile ` is used somewhere in `GenAIComps/.github/workflows/docker/compose`, please fix with the new dir.    


https://github.com/opea-project/GenAIComps/actions/runs/10582938136/job/29344664752?pr=344#step:5:56  
This issue means, those 2 dockerfiles are referred somewhere in GenAIExamples, please submit a PR in GenAIExamples to edit the corresponding place and paste the PR here.

sharanshirodkar7 commented:
@chensuyue Updated `.github/workflows/docker/compose/guardrails-compose.yaml` to reference the correct file path. Also made the corresponding update in the GenAIExamples repository. Here’s the link to the PR: `[GenAIComps/.github/workflows/docker/compose.](https://github.com/opea-project/GenAIExamples/pull/679)`

chensuyue commented:
When merge this PR, please also merge https://github.com/opea-project/GenAIExamples/pull/679.

chensuyue commented:
Why there is some many commit not belong to this PR in the commit list? 
![image](https://github.com/user-attachments/assets/54ad4583-ee02-4559-b434-389954f3b60b)
This make the changed files detect results wrong and return back a large test scope not belong to this PR.

chensuyue commented:
Could you please clean up the commit list and resolve the conflicts?

chensuyue commented on comps/embeddings/predictionguard/README 2.md:
```diff
@@ -0,0 +1,41 @@
+# Embedding Generation Prediction Guard Microservice
```
Why do we need 2 readme?

chensuyue commented on comps/embeddings/predictionguard/requirements 2.txt:
```diff
@@ -0,0 +1,11 @@
+aiohttp
```
Why do we need 2 requirements?

chensuyue commented:
Found a list of duplicate files, please check.

sharanshirodkar7 commented:
@chensuyue 

We have worked on cleaning up the commit list for this PR, including rebasing, removing duplicate files, and resolving conflicts. To keep things organized and clean, we are closing this PR and have opened a new PR here: `https://github.com/opea-project/GenAIComps/pull/605`

Note: This PR has a dependency on the examples repository PR: `https://github.com/opea-project/GenAIExamples/pull/732`
During the process, we encountered several formatting issues flagged by the prettier hook for various files that were not modified by us. We are unsure if it’s our responsibility to address these formatting issues. Could you please provide guidance on whether we need to fix these bugs, or if it would be better handled in a separate PR by the team responsible for those files?

Looking forward to your suggestions.

kevinintel commented:
> sharanshirodkar7

@sharanshirodkar7, it's better to create multiple small PRs instead of a big PR next time.

chensuyue commented:
> @chensuyue
> 
> We have worked on cleaning up the commit list for this PR, including rebasing, removing duplicate files, and resolving conflicts. To keep things organized and clean, we are closing this PR and have opened a new PR here: `https://github.com/opea-project/GenAIComps/pull/605`
> 
> Note: This PR has a dependency on the examples repository PR: `https://github.com/opea-project/GenAIExamples/pull/732` During the process, we encountered several formatting issues flagged by the prettier hook for various files that were not modified by us. We are unsure if it’s our responsibility to address these formatting issues. Could you please provide guidance on whether we need to fix these bugs, or if it would be better handled in a separate PR by the team responsible for those files?
> 
> Looking forward to your suggestions.

You can fix the format issue with local cmd:  
```
pip install pre-commit
pre-commit install
```
After install this tool, when you execute `git commit -m -s xxx`, the format issue will be auto fixed in your PR.

chensuyue commented:
https://github.com/opea-project/GenAIComps/pull/605 Target new PR.
